src/abstractdataframe/sort.jl

```diff
@@ -682,16 +701,91 @@ function Base.sort!(df::AbstractDataFrame, cols=All();
                     lt::Union{Function, AbstractVector{<:Function}}=isless,
                     by::Union{Function, AbstractVector{<:Function}}=identity,
                     rev::Union{Bool, AbstractVector{Bool}}=false,
-                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward)
+                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward,
+                    checkunique::Bool=false)
 
     # exclude AbstractVector as in that case cols can contain order(...) clauses
     if cols isa MultiColumnIndex && !(cols isa AbstractVector)
         cols = index(df)[cols]
     end
     ord = ordering(df, cols, lt, by, rev, order)
     _alg = Sort.defalg(df, ord; alg=alg, cols=cols)
+    checkunique && _perform_uniqueness_checks(df, cols, lt, by, order)
     return sort!(df, _alg, ord)
 end
 
-Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm, o::Base.Sort.Ordering) =
+function Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm,
+                    o::Base.Sort.Ordering)
     permute!(df, _sortperm(df, a, o))
+end
+
+is_complex(o::Ordering) = o isa Union{Order.By, Order.Lt} # Order w/ nonstandard lt or by
```

Comment:
this case of `is_compex` also should check the vale of the field and if it is the default then keep working. Just like below.

Reply:
The relevant code is here https://github.com/JuliaLang/julia/blob/cc7fec740e4349d4036b2771692ad80ef38b975a/base/ordering.jl#L128

Reply:
Again, the problem is that e.g. the following works:

```
julia> Base.Sort.By(identity)
Base.Order.By{typeof(identity), Base.Order.ForwardOrdering}(identity, Base.Order.ForwardOrdering())
```

As with `sort!` case we cannot rely on the fact that currently some type of instance of this object is not created by other functions. The reason is that:
1. user might create it manually.
2. in the future the situation in Base Julia could change.

We need to design code so that it is independent of what other code currently does or does not do. It must be valid in general.

Reply:
Those are very good points. I'll do the checking by hand